drivers/flash: Add flash_get_parameters call to flash API

de-nordic opened the pull request:
Provided call returns const pointer to flash_parameters structure that can be used for passing memory parameters from drivers.

Resolves #22063

de-nordic committed:
drivers/flash: Add API call returning flash parameters

Adds flash_get_parameters call to API that returns pointer to structure
describing flash parameters.
Currently only erase_value parameter is provided via the structure.

Signed-off-by: Dominik Ermel <dominik.ermel@nordicsemi.no>

de-nordic committed:
drivers/flash: Add support for flash_get_parameters to drivers

With addition of flash_get_parameters API call, it is needed to provide
support for the API to flash drivers.

Signed-off-by: Dominik Ermel <dominik.ermel@nordicsemi.no>

de-nordic committed:
drivers/flash: Move write_block_size into flash_parameters

With addition of flash_parameters structure, and supporting API call
to retrieve it, it is no longer needed to store write_block_size as
a part of flash_driver_api and it should be part of flash_parameters.

Signed-off-by: Dominik Ermel <dominik.ermel@nordicsemi.no>

de-nordic committed:
subsys/fs/nvs: Use flash_get_parameters() to get erase value

Use new flash API call to obtain erase value instead of relaying on
hardcoded literals.

Signed-off-by: Dominik Ermel <dominik.ermel@nordicsemi.no>

de-nordic commented on drivers/flash/Kconfig:
```diff
@@ -42,6 +42,33 @@ config FLASH_PAGE_LAYOUT
 	help
 	  Enables API for retrieving the layout of flash memory pages.
 
+choice FLASH_ERASE_BYTE_DEFAULT
+	prompt "Default flash erase value"
+	default FLASH_ERASE_BYTE_FF
+	help
+	  Select value that corresponds to value of erased flash cell.
+
+config FLASH_ERASE_BYTE_FF
+	bool "Erased flash value is 0xFF"
+
+config FLASH_ERASE_BYTE_00
+	bool "Erased flash value is 0x00"
+
+endchoice
```
Not sure whether the selection is needed or the default value should be just hardcoded to 0xff. Please advice.

nvlsianpu replied:
I would resign from FLASH_ERASE_BYTE_VAL - just implement the API for all the driver in the codebase.

nvlsianpu commented on include/drivers/flash.h:
```diff
@@ -296,6 +299,28 @@ static inline size_t z_impl_flash_get_write_block_size(struct device *dev)
+		(const struct flash_driver_api *)dev->driver_api;
+
+	return api->erase_value;
+#else
```
else not needed - introduce this API to all driver instead.

de-nordic replied:
So what you mean is: this is not configurable all drivers fill in the erase_value as an requirement?

nvlsianpu replied:
^^ yes

nvlsianpu replied:
similar as FLASH_PAGE_LAYOUT

de-nordic replied:
OK, if for example the NVS is in use, which requires to know the erase value, then this should be enabled by the dependency, etc?

nvlsianpu reviewed (changes requested):
I like how code looks like now.
Is it possible to make `erased_val` configurable for flash simulator via DTS?
Is it then possible to add test-case with either 0xff and 0x00 erase_value for nvs test (https://github.com/zephyrproject-rtos/zephyr/blob/master/tests/subsys/fs/nvs/testcase.yaml)?

de-nordic commented:
> 
> 
> I like how code looks like now.
> Is it possible to make `erased_val` configurable for flash simulator via DTS?
> Is it then possible to add test-case with either 0xff and 0x00 erase_value for nvs test (https://github.com/zephyrproject-rtos/zephyr/blob/master/tests/subsys/fs/nvs/testcase.yaml)?

Yes. Now the sim flash will have `erase-value` param that allows selection of flash erase value.

nvlsianpu reviewed (approved):
just resolve the conflicts as https://github.com/zephyrproject-rtos/zephyr/pull/25679 imposed badly on nearly all of PRs. :confused:

pabigot commented on include/drivers/flash.h:
```diff
@@ -79,6 +79,9 @@ __subsystem struct flash_driver_api {
 #endif /* CONFIG_FLASH_PAGE_LAYOUT */
 	const size_t write_block_size;
+#if defined(CONFIG_FLASH_ERASE_VALUE)
+	const u8_t erase_value;
```
In theory it may not be possible to determine this value at compile time.  If we generalize QSPI to the point where it supports all flash serial memories that are JESD216 compatible we won't know whether the flash is NOR or NAND until runtime.

I think this should use a function rather than a constant to accommodate that possibility.

pabigot replied:
As a follow-up: can somebody identify a device where the erase value is not 0xFF?  I'm not sure NOR vs NAND produces a difference.

de-nordic replied:
> 
> As a follow-up: can somebody identify a device where the erase value is not 0xFF? I'm not sure NOR vs NAND produces a difference.
> 

No I have failed to identify any FLASH/EEPROM devices that have erase values different from the 0xFF.

utzig replied:
> As a follow-up: can somebody identify a device where the erase value is not 0xFF? I'm not sure NOR vs NAND produces a difference.
> 
> ans: per #22063 "some stm32 embedded"

all STM32L0x and STM32L1x, and AFAIK Cypress PSOC 062/064.

de-nordic replied:
Thanks @utzig for the information, I have been able to go through some of the documents related to the MCUs (ST) you have mentioned. I have just skimmed the documentation, but what seems to be here the case is that the ST MCUs,  does not require erase and have its' flash  accessible via MCU address space the same way as "SRAM" (with proper register setup). So for them the erase_value would not have solved the problem. Some ST expert voice would be useful here.

nvlsianpu replied:
These STM32Lx NVMC performs flash erase operation as well. Most of SoC have access to embedded  flash via memory addresses (nRF as well).  So situuation is allmost the same as for nRF52, Erase value is just different,

pabigot commented on drivers/flash/Kconfig:
```diff
@@ -42,6 +42,13 @@ config FLASH_PAGE_LAYOUT
 	help
 	  Enables API for retrieving the layout of flash memory pages.
 
+config FLASH_ERASE_VALUE
```
Why is this feature conditional?  If drivers can't rely on the API call always being there then they can't guarantee to work on all devices where the erase value might vary.

Also this is misleading in that people will expect this to be the configured erase value, not a flag that enables an API to retrieve the actual erase value.  If it's going to be optional it should be `FLASH_ERASE_VALUE_API`.

nvlsianpu replied:
Generally disabling stuff helps to make footprint smaller - maybe in this case no worth as this will be just a few bytes per a driver instance.

So this can be always available IMO.
Otherwise I think it should be rather default 'y'

pabigot commented on drivers/flash/soc_flash_rv32m1.c:
```diff
@@ -142,6 +142,9 @@ static const struct flash_driver_api flash_mcux_api = {
 	.page_layout = flash_mcux_pages_layout,
 #endif
 	.write_block_size = FSL_FEATURE_FLASH_PFLASH_BLOCK_WRITE_UNIT_SIZE,
+#if IS_ENABLED(CONFIG_FLASH_ERASE_VALUE)
```
For all these: if people accept that my argument that the value needs to support a runtime determination then all these need to move to the device-specific flash or ram structure.

pabigot reviewed (changes requested):
Two major design points to discuss:
* Should the API be available always, or only when specifically enabled;
* Should the implementation support flash devices where we don't know at compile time what the erase value is for a specific device, e.g. determining it from SFDP parameters.

nvlsianpu commented:
@pabigot 
> Should the implementation support flash devices where we don't know at compile time what the erase value is for a specific device, e.g. determining it from SFDP parameters.

Driver implementation might recognize erase-value in the runtime.

pabigot commented:
> Driver implementation might recognize erase-value in the runtime.

Yeah, but then the value can't be stored in the API structure, which is `const` and the same for all devices used by the API.  In that case we need it to be accessed through a function, not by reading the common structure.

nvlsianpu commented:
@pabigot ^^understood, right

Laczen commented:
Maybe the idea of extending the api with a `get_flash_erased_value()` is not the best approach. Maybe it is better/easier to extend the api with a function `flash_iserased(dev, addr, size)` that would do the compare inside the flash driver.

pabigot commented:
> Maybe it is better/easier to extend the api with a function `flash_iserased(dev, addr, size)` that would do the compare inside the flash driver.

It's true that the motivation for this API change has not been documented as clealy as it could be.

There is a use for initializing a buffer with the intent that any elements not specifically assigned otherwise will not mutate the content of the flash, essentially preserving the unerased parts.  That's currently handled in one situation by first reading the region, modifying parts, and writing the whole thing back.

Is that an approach that would solve #22063 without changing the flash API?

Even in that case knowing the erase value would be necessary to verify that the technique won't break (e.g. flash already has 0xAA instead of 0xFF and user wants to write 0xF3 to set bits 0..3, but the result would be 0xA2 which has the wrong value in bits 0..3).

Laczen commented:
@pabigot, I think this case is different, nvs is searching for the end of the fs by searching for a unwritten value. The read and rewrite to update the flash data will not help.
@pabigot, regarding the read and rewrite for mcuboot support, it might be a bad idea to support this in general, the result might be that flash wear is a lot faster than expected or the method is not supported at all by the hardware. Mcuboot afaik always states that they do not need byte access write, if mcuboot does need read and write back it would be best to add this in mcuboot.

de-nordic commented:
@Laczen @nvlsianpu @pabigot I have search for information on flash and eeproms to find if there is device that does erase to non-1 value but I have failed to identify any. 
There are FRAM (FeRAM) devices and MRAM, but they behave like RAM generally which means that they do not require erase prior to write.

Maybe this PR should be dropped. I can not find an example of non-1 erase value.
OK, so NVS would not work with some other erase values: yes but we do not have example of such.
In case of FRAM and MRAM, the NVS will not work on them either because they do not have "erase value", support for them  could be emulated via flash API if we add erase functions for them that would default the RAM to all 1 (which will cost us some time).

nvlsianpu commented:
> Mcuboot afaik always states that they do not need byte access write, if mcuboot does need read and write back it would be best to add this in mcuboot.

Mcuboot doesn't and won't use any read-modify-write to flash. It just need to know what the erased value is for working properly on SoC of 0x00 and 0xFF erased values

nvlsianpu commented:
@de-nordic +1 for introducing this API.
cc @utzig

de-nordic commented:
> 
> 
> @de-nordic +1 for introducing this API.
> cc @utzig

@nvlsianpu  I still voice my concern regarding other technologies of NV type memories, but I am modifying the PR to implement @pabigot suggestions regarding the callback instead of data member.  Additionally I will change it from erase specific callback to `flash_get_capabilities`, `flash_get_params` or similar, that could be used in more general way to obtain flash parameters that could be defined during compilation or run-time.

I am also looking into NVS driver to figure out how would it be possible to make it work with non-volatile memories that does not require erase prior to write.

carlescufi commented:
@pabigot can you re-review please?

nvlsianpu commented:
Maybe for memories which have not defined erase-value (or erase is not an operation for them) API can return -ENOSUP etc.

de-nordic commented:
> 
> 
> @pabigot can you re-review please?

There are no changes yet. I am still working o it. @pabigot I will let you know when I will have something to present, although I would like to hear your point of view on current support of NV type memories in Zephyr and whether flash/eeprom api shouldn't be merged under one NV API that would also cover capabilities of other type of NV memories.
Actually I would also encourage @nvlsianpu  and @Laczen, to voice their opinions on the subject as they are more experienced, than me, with the Zephyr API and NV memory usage within the system.
Thanks in advance!

henrikbrixandersen commented:
> I think it's something to consider but @henrikbrixandersen initiated that API and should weigh in on whether its operations can be supported with the flash API. I assume in that situation we'd need to define the expected behavior of erase: should it do nothing, or should it write a constant to remove any information present (I would propose the latter).

We had the discussion of whether to use the flash API for EEPROM devices (and whether to add an erase API function) back when the EEPROM API was introduced. Please see #19972 for details.

nvlsianpu commented on tests/drivers/flash_simulator/prj.conf:
```diff
@@ -1,5 +1,6 @@
 CONFIG_ZTEST=y
 CONFIG_FLASH=y
+CONFIG_FLASH_ERASE_VALUE=y
```
```suggestion
CONFIG_FLASH_GET_PARAMETERS_API=y
```

nvlsianpu replied:
fix in proper commit

pabigot replied:
Agree, this doesn't belong in this commit.

hakonfam commented on drivers/flash/Kconfig:
```diff
@@ -42,6 +42,13 @@ config FLASH_PAGE_LAYOUT
 	  Enables API for retrieving the layout of flash memory pages.
 
+config FLASH_GET_PARAMETERS_API
+	bool "Flash parameters are avaialble via flash_get_parameters"
```
```suggestion
	bool "Flash parameters are available via flash_get_parameters"
```

nvlsianpu replied:
not configurable anymore

de-nordic replied:
Yes it has been thrown out.

hakonfam commented on include/drivers/flash.h:
```diff
@@ -36,12 +36,20 @@ struct flash_pages_layout {
+ * Flash memory parameters
+ */
+struct flash_parameters {
+	uint8_t erase_value; /* Pattern representing of erased flash */
```
I guess it should be either
`Pattern representation of erased flash`
or
`Pattern representing erased flash` 
?

pabigot replied:
Or "byte value read back from erased area"

hakonfam reviewed (commented):
some nits

pabigot commented on include/drivers/flash.h:
```diff
@@ -36,12 +36,20 @@ struct flash_pages_layout {
+ * Flash memory parameters
+ */
+struct flash_parameters {
+	uint8_t erase_value; /* Pattern representing erased flash */
```
Or "Byte value of erased flash".  Not terribly important, but "pattern" to me suggests a mask rather than a value.

pabigot commented on include/drivers/flash.h:
```diff
@@ -296,6 +307,25 @@ static inline size_t z_impl_flash_get_write_block_size(struct device *dev)
+/**
+ *  @brief  Get pointer to flash_parameters structure
+ *
+ *  @return pointer to flash_parameters structure characteristic for
```
... or NULL if parameters are not available.

de-nordic replied:
No NULL, withdrawing from making this optional.

pabigot commented on include/drivers/flash.h:
```diff
@@ -296,6 +307,25 @@ static inline size_t z_impl_flash_get_write_block_size(struct device *dev)
+	const struct flash_driver_api *api =
+		(const struct flash_driver_api *)dev->driver_api;
+
+	return api->get_parameters();
```
Because this API is optional, it'd be easy for out-of-tree drivers to not provide it, and never notice that somebody enabled the option.  There should at minimum be an `__ASSERT()` that the API field has been filled out, but I'd still rather see:

    return (api->get_parameters == NULL) ? NULL : api->get_parameters();

to at least let the application be able to recover in that situation.

Which then gets back to the question of whether `flash_get_parameters()` should always be available, and just return `NULL` if the option isn't enabled.  I don't recall what the resolution of that discussion was, if we reached one.

de-nordic replied:
I will make it permanent addition to API. I think it is not worth making this optional as checking whether this returns  NULL, after calls, is too much of a hassle.

de-nordic replied:
Even if turning this off saves us ~8 bytes of flash, the checking for NULL every time when the structure is requested would probably cost us as much more. And NULL processing at runtime is problematic because this is a thing that  is supposed to describe some property, what if it is not there? Stop running?

pabigot commented on tests/drivers/flash_simulator/src/main.c:
```diff
@@ -266,6 +269,15 @@ static void test_double_write(void)
+	const struct flash_parameters *fp = flash_get_parameters(flash_dev);
+
+	zassert_equal(fp->erase_value, FLASH_SIMULATOR_ERASE_VALUE,
+		      "Eipected erase value %hhx",
```
s/Eipected/Expected/.

Also `%x` not `%hhx`.  The value gets promoted to unsigned int in the call, no need for unusual length modifiers.

pabigot reviewed (changes requested):
I think I'd like the API to return `NULL` when the parameters are not available, and allow drivers to not populate the implementation pointer in that case.

Alternatively if all drivers must implement this API, then there needs to be an `__ASSERT()` in the impl wrapper so that drivers that don't set the field don't call through a null pointer.

pabigot commented on include/drivers/flash.h:
```diff
@@ -79,6 +87,7 @@ __subsystem struct flash_driver_api {
 	flash_api_pages_layout page_layout;
 #endif /* CONFIG_FLASH_PAGE_LAYOUT */
 	const size_t write_block_size;
+	flash_api_get_parameters get_parameters;
```
OK, so since we're adding API unconditionally, this raises new questions.

`write_block_size` is clearly a parameter of the flash.  Shouldn't it move to this structure?

Then we can deprecate `flash_get_write_block_size()` and implement it temporarily using `flash_get_parameters()`.

I realize this starts down a slippery slope, but it looks like maybe we should start sliding.

de-nordic replied:
I have been thinking about moving the `write_block_size` to flash_parameters, though I am not sure whether this, as it would also men depreciation of existing API, in this PR.
@pabigot Should we do that now? Why is it slippery slope?

pabigot replied:
Well, we don't *have* to deprecate the API, we can just implement in terms of the new API and leave it.  But it could be deprecated since it'd be redundant with the value in the parameter structure.

"Slippery" because we should also look at what's in the page layout description and see what else should be exposed.  Like whether there's a fixed minimum erase size.

I think that should be left for a separate PR, but that moving `write_block_size` is appropriate as it justifies the move to an aggregate parameter structure (we had just one value, now we have two, so we should generalize).

de-nordic replied:
OK, I will mote the `write_block_size` into the `flash_parameters` structure within this PR.

pabigot commented on include/drivers/flash.h:
```diff
@@ -36,12 +36,20 @@ struct flash_pages_layout {
 			       const void *data, size_t len);
 typedef int (*flash_api_erase)(struct device *dev, off_t offset, size_t size);
 typedef int (*flash_api_write_protection)(struct device *dev, bool enable);
+typedef const struct flash_parameters* (*flash_api_get_parameters)(void);
```
Thinking again: this needs to take a `struct device *dev` parameter.  There's an outside chance of a generic flash driver for which different devices have different properties.

If we were not to do that (and I really do think we should) we could instead just store the pointer to the structure in the API structure, as is currently done for `write_block_size`.

de-nordic replied:
OK, I will change this.

pabigot reviewed (changes requested):
Looks like this is one of those things that looks easy but creates more questions every time we iterate.

In this case I notice we have an existing flash property that should go into the new structure, unless there's a good reason why it shouldn't.  Also we should support having different parameters for different flash devices.

pabigot commented on drivers/flash/flash_gecko.c:
```diff
@@ -178,6 +184,14 @@ void flash_gecko_page_layout(struct device *dev,
+{
+	ARG_UNUSED(dev);
+
+	return (const struct flash_parameters *)&flash_gecko_parameters;
```
The cast of the address is unnecessary and confusing.  Please remove from all drivers.

de-nordic replied:
Yes. I do not even remember anymore, why I am doing this casting...

pabigot commented on drivers/flash/flash_sam0.c:
```diff
@@ -417,11 +434,7 @@ static const struct flash_driver_api flash_sam0_api = {
-#else
-	.write_block_size = FLASH_PAGE_SIZE,
-#endif
+	.get_parameters = flash_sam0_get_parameters,
```
I have a slight preference for having the unconditional fields appear before any conditional fields, so would put this before `.page_layout`.  Just a suggestion.

de-nordic replied:
My personal preference is to initialize fields, if possible, in the order of appearance. This makes easier for me to figure out if I have missed something, etc. Looks messier when a lot of c preprocessor is involved but still easier to figure out. Do you want me to change it?

pabigot replied:
Initializing in the order of definition isn't just a preference; it's a requirement for C++.  What I meant was to move the `get_parameters` definition up next to the other ones, and keep the optional ones at the end.  Since it's new, it doesn't matter whether it goes at the end or is inserted: the relative order of everything else remains unchanged.

pabigot replied:
> Do you want me to change it?

Yes, but I'm not going to block if you don't.

de-nordic replied:
OK I will do this.

de-nordic replied:
For C99 the order of definition of structure fields, when using "designators", does not matter.

pabigot commented on drivers/flash/flash_stm32.c:
```diff
@@ -57,6 +57,18 @@ LOG_MODULE_REGISTER(flash_stm32, CONFIG_FLASH_LOG_LEVEL);
 #define CFG_HW_FLASH_SEMID	2
 
+static const struct flash_parameters flash_stm32_parameters = {
+#endif
```
This looks out of place.

de-nordic replied:
Fixed.

pabigot commented on drivers/flash/spi_nor.c:
```diff
@@ -84,6 +84,12 @@ struct spi_nor_data {
 	struct k_sem sem;
 };
 
+
```
unexpected blank line

de-nordic replied:
Fixed

pabigot commented on drivers/flash/nrf_qspi_nor.c:
```diff
@@ -33,6 +33,10 @@
 LOG_MODULE_REGISTER(qspi_nor, CONFIG_FLASH_LOG_LEVEL);
 
+static const struct flash_parameters qspi_flash_parameters = {
+	.erase_value = 0xff,
```
missing `.write_block_size`.  Multiple drivers.

de-nordic replied:
Eh. Failed to properly apply patches between branches...

pabigot reviewed (commented):
More preliminary feedback; many drivers need to be fixed to move the `write_block_size` from the API structure to the parameters.

Laczen commented:
@de-nordic, it seems that the flash api extension is getting defined.

IMO the best way to use this extension in nvs is to extend the nvs fs structure with a pointer to the flash parameters and to set this during nvs_init. If any of the required flash parameters is missing raise a cfg error and keep the nvs flash_device = null. In the nvs routines use the nvs pointer to flash parameters  (so the get_flash_param() is only used in nvs_init and not inside any other nvs routines).

de-nordic commented:
> 
> IMO the best way to use this extension in nvs is to extend the nvs fs structure with a pointer to the flash parameters and to set this during nvs_init. If any of the required flash parameters is missing raise a cfg error and keep the nvs flash_device = null. In the nvs routines use the nvs pointer to flash parameters (so the get_flash_param() is only used in nvs_init and not inside any other nvs routines).

I will look into this.

de-nordic committed:
subsys/fs/nvs: Move write_block_size to flash_parameters

Pointer to flash_parameters have been added to nvs_fs structure and it
is no longer needed to store write_block_size within the nvs_fs.

Signed-off-by: Dominik Ermel <dominik.ermel@nordicsemi.no>

de-nordic commented:
@Laczen does the changes to NVS follow your idea?

Laczen commented:
> @Laczen does the changes to NVS follow your idea?

Yes, this is completely what I was thinking. I don't know if it is possible to use a `const struct *flash_parameters`, it might be easier to use when it's not `const` and gets assigned during init (and a check is made on the write_block_size).

de-nordic commented:
> 
> Yes, this is completely what I was thinking. I don't know if it is possible to use a `const struct *flash_parameters`.

@Laczen  How do you want to use it? I do not encourage anyone to modify contents of `flash_parameters` outside of drivers.

pabigot commented:
> Yes, this is completely what I was thinking. I don't know if it is possible to use a `const struct *flash_parameters`.

That should be easy to do: you return a pointer to const, but keep the structure in RAM and initialize it when the device is initialized.

I agree the API should not provide a pointer to non-const.  I'd like to document that the caller has the choice of caching the pointer, or copying the information desired to a different data structure with a guarantee it will never change.

Laczen commented:
> > Yes, this is completely what I was thinking. I don't know if it is possible to use a `const struct *flash_parameters`.
> 
> @Laczen How do you want to use it? I do not encourage anyone to modify contents of `flash_parameters` outside of drivers.

Exactly how you are using it. Probably my remark is a result of limited c knowledge, can you assign the `const struct *flash_parameters` in nvs_init() ?

de-nordic commented:
> 
> 
> > > Yes, this is completely what I was thinking. I don't know if it is possible to use a `const struct *flash_parameters`.
> > 
> > 
> > @Laczen How do you want to use it? I do not encourage anyone to modify contents of `flash_parameters` outside of drivers.
> 
> Exactly how you are using it. Probably my remark is a result of limited c knowledge, can you assign the `const struct *flash_parameters` in nvs_init() ?

Yes, `const` qualifier 'protects' here the structure as it is in the part that declares pointed type. If you want to protect pointer, you need to move `const`:
`struct *const flash_parameters`
or you can have both:
`const struct *const flash_parameters`

de-nordic commented:
@pabigot 

> 
> 
>(...) I'd like to document that the caller has the choice of caching the pointer, or copying the information desired to a different data structure with a guarantee it will never change.

You want this in the doxygen of the `flash_get_parameters()` ?

pabigot commented:
> > (...) I'd like to document that the caller has the choice of caching the pointer, or copying the information desired to a different data structure with a guarantee it will never change.
> 
> You want this in the doxygen of the `flash_get_parameters()` ?

I think so, yes, assuming you and others think this is worth making clear.  I'd like to have this API raised during the API telecon today for visibility.  But also we should confirm that nobody has a use case for the values changing for a given device while the system is running.

de-nordic commented:
> 
> I think so, yes, assuming you and others think this is worth making clear. I'd like to have this API raised during the API telecon today for visibility. But also we should confirm that nobody has a use case for the values changing for a given device while the system is running.

I do not think there is a case for the contents of `flash_parameters` to get modified at runtime. I will add the note to documentation, but we will need to wait until it gets clarified.

Laczen commented:
> > > > Yes, this is completely what I was thinking. I don't know if it is possible to use a `const struct *flash_parameters`.
> > > 
> > > 
> > > @Laczen How do you want to use it? I do not encourage anyone to modify contents of `flash_parameters` outside of drivers.
> > 
> > 
> > Exactly how you are using it. Probably my remark is a result of limited c knowledge, can you assign the `const struct *flash_parameters` in nvs_init() ?
> 
> Yes, `const` qualifier 'protects' here the structure as it is in the part that declares pointed type. If you want to protect pointer, you need to move `const`:
> `struct *const flash_parameters`
> or you can have both:
> `const struct *const flash_parameters`

OK, thanks for the clarification

nvlsianpu commented:
NVS changes looks ok

erwango commented on drivers/flash/flash_stm32.c:
```diff
@@ -57,6 +57,10 @@ LOG_MODULE_REGISTER(flash_stm32, CONFIG_FLASH_LOG_LEVEL);
 #define CFG_HW_FLASH_SEMID	2
 
+static const struct flash_parameters flash_stm32_parameters = {
+	.erase_value = 0xff,
```
This doesn't apply to all series.
On a handfull, this should be 0x00.
I'll get back to you with the references

de-nordic replied:
OK, thanks for looking into this.

de-nordic replied:
@utzig has mentioned same thing here: https://github.com/zephyrproject-rtos/zephyr/pull/25947#discussion_r438000040 though I do not know how does it apply to our stm32 drivers.

de-nordic replied:
@erwango Can you advise me on the stm32 characteristic values for `erase_value`?

erwango replied:
@de-nordic After check, only L0/L1 use 0x00 as erase value (I guess reason is they have an EEPROM and this is easier to use same erase value for both).
Since L0/L1 have no flash driver yet, proposed code is fine. Maybe adding a warning comment in case of L0/L1 future addition would be a nice thing though.

de-nordic replied:
@erwango Thanks for the information. Will add warning information.

andysan replied:
There is a PR for an L0 driver (https://github.com/zephyrproject-rtos/zephyr/pull/25245) that will hit this issue. The easiest way to distinguish between the EEPROM and FLASH implementations is probably to check for the FLASH_CR_LOCK (erase value 0xff) or FLASH_PECR_PELOCK (erase value 0x00). It's not exactly pretty, but it avoids hard-coding the SOC family.

Laczen replied:
> There is a PR for an L0 driver (#25245) that will hit this issue. The easiest way to distinguish between the EEPROM and FLASH implementations is probably to check for the FLASH_CR_LOCK (erase value 0xff) or FLASH_PECR_PELOCK (erase value 0x00). It's not exactly pretty, but it avoids hard-coding the SOC family.

@andysan, does this mean that also for L0 the flash erase value is 0xff?

carlescufi commented:
@pabigot @nvlsianpu could you take another look please?

pabigot commented on drivers/flash/flash_simulator.c:
```diff
@@ -343,15 +350,23 @@ static void flash_sim_page_layout(struct device *dev,
 #ifdef CONFIG_FLASH_PAGE_LAYOUT
 	.page_layout = flash_sim_page_layout,
 #endif
+	.get_parameters = flash_sim_get_parameters,
```
Placement of this one isn't the same as the others.

de-nordic replied:
Done

pabigot commented on include/drivers/flash.h:
```diff
@@ -293,7 +304,30 @@ static inline size_t z_impl_flash_get_write_block_size(struct device *dev)
+ *  constant through a runtime, regardless if it is defined in RAM or
+ *  Flash.
+ *  Developer is free to cache the structure pointer or copy its contents,
+ *  although a pointer if preferred method of access.
```
I'd make this end at "copy its contents."  The rest is opinion that isn't good advice for some uses.  If you want it anyway just change "if preferred" to "is the preferred".

de-nordic replied:
Done

pabigot reviewed (approved):
OK as is, a couple minor cleanups identified.

de-nordic commented:
@Laczen @pabigot Thanks for reviewing.
